mjpc/planners/sample_gradient/planner.cc

```diff
@@ -0,0 +1,652 @@
+
+  // ----- update policy ----- //
+  // start timer
+  auto policy_update_start = std::chrono::steady_clock::now();
+
+  // sort candidate policies and trajectories by score
+  for (int i = 0; i < num_trajectory; i++) {
+    trajectory_order[i] = i;
+  }
+
+  // sort so that the first ncandidates elements are the best candidates, and
+  // the rest are in an unspecified order
+  std::partial_sort(
+      trajectory_order.begin(), trajectory_order.begin() + num_trajectory,
+      trajectory_order.begin() + num_trajectory,
+      [&trajectory = trajectory](int a, int b) {
+        return trajectory[a].total_return < trajectory[b].total_return;
+      });
+
+  // set winner
+  if (trajectory[trajectory_order[0]].total_return <
+      trajectory[idx_nominal].total_return) {
+    winner = trajectory_order[0];
+  } else {
+    winner = idx_nominal;
+  }
+
+  // winner type
+  if (winner > idx_nominal) {
+    if (winner < num_trajectory - num_gradient) {
+      winner_type_ = kPerturb;
+    } else {
+      winner_type_ = kGradient;
+    }
+  } else {
+    winner_type_ = kNominal;
+  }
+
+  // update
+  {
+    const std::shared_lock<std::shared_mutex> lock(mtx_);
+    policy.CopyParametersFrom(candidate_policy[winner].parameters,
+                              times_scratch);
+  }
+
+  // improvement: compare nominal to elite average
+  improvement = mju_max(
+      trajectory[idx_nominal].total_return - trajectory[winner].total_return,
+      0.0);
+
+  // stop timer
+  policy_update_compute_time = GetDuration(policy_update_start);
+
+  // ----- compute gradient candidate policies ----- //
+  // start timer
+  auto gradient_start = std::chrono::steady_clock::now();
+
+  // candidate policies
+  this->GradientCandidates(num_trajectory, num_gradient, horizon, pool);
+
+  // stop timer
+  gradient_candidates_compute_time = GetDuration(gradient_start);
+}
+
+// compute trajectory using nominal policy
+void SampleGradientPlanner::NominalTrajectory(int horizon, ThreadPool& pool) {
+  // set policy
+  auto nominal_policy = [&cp = resampled_policy](
+                            double* action, const double* state, double time) {
+    cp.Action(action, state, time);
+  };
+
+  // rollout nominal policy
+  trajectory[idx_nominal].Rollout(nominal_policy, task, model, data_[0].get(),
+                                  state.data(), time, mocap.data(),
+                                  userdata.data(), horizon);
+}
+
+// set action from policy
+void SampleGradientPlanner::ActionFromPolicy(double* action,
+                                             const double* state, double time,
+                                             bool use_previous) {
+  const std::shared_lock<std::shared_mutex> lock(mtx_);
+  if (use_previous) {
+    previous_policy.Action(action, state, time);
+  } else {
+    policy.Action(action, state, time);
+  }
+}
+
+// update policy via resampling
+void SampleGradientPlanner::ResamplePolicy(
```

Comment:
Maybe too much hassle, but would it be better to inherit this?

@nimrod-gileadi is making improvements to regular sampling that this class would also enjoy. Nimrod WDYT?

Reply:
This routine is modified from `SamplingPlanner` to enable any of the policies to be resampled. We could inherit if we use this version with that planner.